bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsServlet.java

```diff
@@ -0,0 +1,174 @@
+        if (httpService == null) {
+            logger.warn("SmartthingsServlet.activate: httpService is unexpectedly null");
+            return;
+        }
+        try {
+            Dictionary<String, String> servletParams = new Hashtable<String, String>();
+            httpService.registerServlet(PATH, this, servletParams, httpService.createDefaultHttpContext());
+        } catch (ServletException | NamespaceException e) {
+            logger.warn("Could not start Smartthings servlet service: {}", e.getMessage());
+        }
+    }
+
+    @Deactivate
+    protected void deactivate(ComponentContext componentContext) {
+        if (httpService != null) {
+            try {
+                httpService.unregister(PATH);
+            } catch (IllegalArgumentException ignored) {
+            }
+        }
+    }
+
+    @Reference
+    protected void setHttpService(HttpService httpService) {
+        this.httpService = httpService;
+    }
+
+    protected void unsetHttpService(HttpService httpService) {
+        this.httpService = null;
+    }
+
+    @Reference
+    protected void setEventAdmin(EventAdmin eventAdmin) {
+        this.eventAdmin = eventAdmin;
+    }
+
+    protected void unsetEventAdmin(EventAdmin eventAdmin) {
+        this.eventAdmin = null;
+    }
+
+    @Override
+    protected void service(@Nullable HttpServletRequest req, @Nullable HttpServletResponse resp)
+            throws ServletException, IOException {
+        if (req == null) {
+            logger.debug("SmartthingsServlet.service unexpectedly received a null request. Request not processed");
+            return;
+        }
+        String path = req.getRequestURI();
+
+        // See what is in the path
+        String[] pathParts = path.replace(PATH + "/", "").split("/");
+        if (pathParts.length != 1) {
+            logger.warn(
+                    "Smartthing servlet received a path with zero or more than one parts. Only one part is allowed. path {}",
+                    path);
+            return;
+        }
+
+        BufferedReader rdr = new BufferedReader(req.getReader());
+        String s = rdr.lines().collect(Collectors.joining());
+        switch (pathParts[0]) {
+            case "state":
+                // This is device state info returned from Smartthings
+                logger.debug("Smartthing servlet processing \"state\" request. data: {}", s);
+                publishEvent(STATE_EVENT_TOPIC, "data", s);
+                break;
+            case "discovery":
+                // This is discovery data returned from Smartthings
+                logger.trace("Smartthing servlet processing \"discovery\" request. data: {}", s);
+                publishEvent(DISCOVERY_EVENT_TOPIC, "data", s);
+                break;
+            case "error":
+                // This is an error message from smartthings
+                Map<String, Object> map = new HashMap<String, Object>();
+                map = gson.fromJson(s, map.getClass());
```

Comment:
Does gson allow you to convert json into a `Map<String, Object>`? How would that even work, what class is it using for the `Object` values?

Reply:
Yes this did work. The value (Object type) is a String. But, I changed it to <String, String>